docs/proposals/auth.md

```diff
@@ -0,0 +1,150 @@
+    AuthorizeCreation(user kauthuser.Info, keys []MetricKey, prefix string, metricNames []string) error 
+
+    // AuthorizeQuery authorizes querying of metrics at the given keys with the
+    // given names from the given source.
+    AuthorizeQuery(user kauthuser.Info, source HeapsterQuerySource, keys []MetricKey, metricNames []string) error
+}
+```
+
+The `Authorizer` interface allows authorizing both queries against Heapster
+and requests to create new metrics entries.
+
+For queries, `AuthorizeQuery` takes user information, a source from which the
+information will be queried (the model or the historical source), and a set of
+keys and metric names.
+
+The authorizer may not assume a relationship between keys and metric names --
+it must assume all metric names given may be pushed for all metric keys (see
+below for the rationale for passing metric names).
+
+The authorizer will then take the given information, and return an error if
+the given user is not authorized to perform the given query.
+
+For creation of metrics, `AuthorizeCreation` takes much the same parameters
+as `AuthorizeQuery` (except for the `source` parameter), with the addition
+of the `prefix` parameter.  If the prefix parameter is given, when checking
+metric names, the authorizer should first prepend the prefix to the metric
+names.
+
+Passing keys to the authorizer allows for fine-grained scope decisions to be
+made -- some users may be only able to query metrics in their own namespace,
+while others may be able to query any metrics.  Some users may be able to push
+metrics only for their own namespace, or even for their own pod, while others
+may be able to push metrics for any key in Heapster.
+
+Passing in metric names allows the authorizer to make complex decisions about
+which users are allowed to push all which metrics.  For instance, an authorizer
+could restrict certain users to pushing metrics with a certain prefix, while
+allowing another user to push metrics without a prefix, as long as they were
+not one of the built-in metrics.  On the query side, this could allow
+metric-level restrictions on queries, such as only allowing certain users to
+read a certain set of custom metrics, but not have access to built-in metrics.
+
+If the `keys` parameter is emtpy, the authorizer should treat the request as
+if the user were requesting access to all keys.  If the `names` field is empty,
+the authorizer should treat the request as if the user were requesting access
+to all names prefixed with the prefix given by `prefix`.
+
+To replicate the current client certificate mechanism, an implementation would
+simply accept any request (optionally checking the user name against a list of
+allowed users, as in the current setup).
+
+A system based on OpenShift/Kube authorization might generate a
+`SubjectAccessReview` request against an OpenShift or Kube server (e.g.
+mapping ability to `GET` pods to the ability to query the model).
```

Comment:
If just Heapster does this, it is not so bad maybe, but if lots of peripheral systems to this, it makes the meaning of "GET pods" encompass a lot of things, which makes it hard for someone to understand this permission.  

Once you have a fairly general IAM system like RBAC, you might as well express the permissions within the RBAC framework.  And then extend the default roles to have the appropriate permissions.

This also allows you to show a users permissions explicitly in a clear way in a CLI/GUI.  (e.g. when you list "roleFoo" permissions, you see both "GET pods in this namespace" and "GET modelMetrics in this namespace" and "GET historicalMetrics in this namespace".